autorest/azure/auth/auth.go

```diff
@@ -41,59 +41,105 @@ import (
 // 3. Username password
 // 4. MSI
 func NewAuthorizerFromEnvironment() (autorest.Authorizer, error) {
-	tenantID := os.Getenv("AZURE_TENANT_ID")
-	clientID := os.Getenv("AZURE_CLIENT_ID")
-	clientSecret := os.Getenv("AZURE_CLIENT_SECRET")
-	certificatePath := os.Getenv("AZURE_CERTIFICATE_PATH")
-	certificatePassword := os.Getenv("AZURE_CERTIFICATE_PASSWORD")
-	username := os.Getenv("AZURE_USERNAME")
-	password := os.Getenv("AZURE_PASSWORD")
-	envName := os.Getenv("AZURE_ENVIRONMENT")
-	resource := os.Getenv("AZURE_AD_RESOURCE")
-
-	var env azure.Environment
+	settings := getEnvVars()
+
+	env, err := getEnv(settings.envName)
+	if err != nil {
+		return nil, err
+	}
+
+	if settings.resource == "" {
+		settings.resource = env.ResourceManagerEndpoint
+	}
+
+	return getAuth(settings, env)
+}
+
+// NewKeyvaultAuthorizerFromEnvironment creates a keyvault dataplane Authorizer configured from environment variables in the order:
+// 1. Client credentials
+// 2. Client certificate
+// 3. Username password
+// 4. MSI
+func NewKeyvaultAuthorizerFromEnvironment() (autorest.Authorizer, error) {
+	settings := getEnvVars()
+	settings.resource = os.Getenv("AZURE_KEYVAULT_RESOURCE")
+
+	env, err := getEnv(settings.envName)
+	if err != nil {
+		return nil, err
+	}
+
+	if settings.resource == "" {
+		settings.resource = strings.TrimSuffix(env.KeyVaultEndpoint, "/")
+	}
+
+	return getAuth(settings, env)
+}
+
+type envSettings struct {
```

Comment:
I'd like to change this name. I think it'll be easily confused with this: https://github.com/Azure/go-autorest/blob/1ff28809256a84bb6966640ff3d0371af82ccba4/autorest/azure/environments.go#L37

Not to mention, I think it also implies that the only way to populate these are with Environment variables. (Which totally is one good way, but maybe we want expose others someday. Command line args parser, for example.)

Maybe `identityKeys`, `authKeys` or `authSettings` would be better?

Reply:
Fixed